File: Archived_Code_20200906_1604/Backup/Code_17122019_Backup_Do_Not_Delete/mount-everest/Sorting/merge-sort.cc
#include <iostream>
#include <ctime>
#include <vector>
#include <algorithm>

std::vector<int> generateRandomNumbers(int n, int min_num, int max_num) {
    std::vector<int> nums(n);
    for (int i = 0; i< n; i++) {
        nums[i] = min_num + rand() % (max_num - min_num + 1); 
    }
    return nums;
}

void print(int arr[], int n) {
    for (int i = 0; i < n; i++) {
        std::cout << arr[i] << " ";
    }
    std::cout << "\n";
}

void merge(int arr[], int start, int mid, int end) {
    int p = start, q = mid + 1;
    int temp[end - start + 1], k = 0;
    for (int i = start; i <= end; i++) {
        if (p > mid) {
            temp[k++] = arr[q++];
        } else if (q > end) {
            temp[k++] = arr[p++];
        } else if (arr[p] < arr[q]) {
            temp[k++] = arr[p++];
        } else {
            temp[k++] = arr[q++];
        }
    }

    for (int p = 0; p < k; p++) {
        arr[start++] = temp[p];
    }
}

void merge_sort(int arr[], int l, int r) {
    if (l < r) {
        int m = l + (r - l) / 2;
        merge_sort(arr, l, m);
        merge_sort(arr, m+1, r);
        merge(arr, l, m, r);
    }
}

int main (int argc, char *argv[]) {
    srand(time(nullptr));
    std::vector<int> nums = generateRandomNumbers(20, 10, 99);
    int *arr = nums.data();
    print(arr, 20);
    merge_sort(arr, 0, nums.size() - 1);
    print(arr, 20);

    return 0;
}

File: 20210816-0947/Archived_Code_20200906_1604/Backup/Code_17122019_Backup_Do_Not_Delete/java-revisited/src/Sorting/MergeSort.java
package Sorting;

import java.util.Arrays;

public class MergeSort {

    public static void merge(int[] array, int start, int mid, int end) {
        int p = start, q = mid + 1;
        int k = 0;
        int[] arr = new int[end - start + 1];
        for (int i = start; i <= end; i++) {
            if (p > mid)
                arr[k++] = array[q++];
            else if (q > end)
                arr[k++] = array[p++];
            else if (array[p] < array[q])
                arr[k++] = array[p++];
            else
                arr[k++] = array[q++];
        }

        for (int i = 0; i < k; i++) {
            array[start++] = arr[i];
        }
    }

    public static void merge_sort(int[] array, int start, int end) {
        if (start < end) {
            int mid = (start + end) / 2;
            merge_sort(array, start, mid);
            merge_sort(array, mid + 1, end);
            merge(array, start, mid, end);
        }
    }

    public static void main(String[] args) {
        int[] array = {14, 20, 24, 11, 16};
        System.out.println("Before sorting = " + Arrays.toString(array));
        merge_sort(array, 0, array.length - 1);
        System.out.println("After sorting = " + Arrays.toString(array));
    }
}
